community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulator.java

```diff
@@ -0,0 +1,418 @@
+     * <p>
+     * <b>NOTE:</b> for testing only.
+     *
+     * @param storeView the view of the store as a visitable of nodes
+     * @param executor the thread pool to use for batched index insertions
+     * @param logProvider the log provider
+     */
+    BatchingMultipleIndexPopulator( IndexStoreView storeView, ExecutorService executor, LogProvider logProvider )
+    {
+        super( storeView, logProvider );
+        this.executor = executor;
+    }
+
+    @Override
+    public StoreScan<IndexPopulationFailedKernelException> indexAllNodes()
+    {
+        StoreScan<IndexPopulationFailedKernelException> storeScan = super.indexAllNodes();
+        return new BatchingStoreScan<>( storeScan );
+    }
+
+    @Override
+    protected IndexPopulation createPopulation( IndexPopulator populator,
+            IndexDescriptor descriptor, IndexConfiguration config, SchemaIndexProvider.Descriptor providerDescriptor,
+            FlippableIndexProxy flipper, FailedIndexProxyFactory failedIndexProxyFactory, String indexUserDescription )
+    {
+        return new BatchingIndexPopulation( populator, descriptor, config, providerDescriptor, flipper,
+                failedIndexProxyFactory, indexUserDescription );
+    }
+
+    @Override
+    protected void populateFromQueue( long currentlyIndexedNodeId )
+    {
+        if ( queue.size() >= QUEUE_THRESHOLD )
+        {
+            log.debug( "Populating from queue." + EOL + this );
+            flushAll();
+            awaitCompletion();
+            super.populateFromQueue( currentlyIndexedNodeId );
+            log.debug( "Drained queue and all batched updates." + EOL + this );
+        }
+    }
+
+    @Override
+    public String toString()
+    {
+        String updatesString = batchedUpdates.entrySet()
+                .stream()
+                .map( entry -> entry.getKey() + " - " + entry.getValue().size() + " updates" )
+                .collect( joining( ", ", "[", "]" ) );
+
+        return "BatchingMultipleIndexPopulator{activeTasks=" + activeTasks + ", executor=" + executor + ", " +
+               "batchedUpdates = " + updatesString + ", queuedUpdates = " + queue.size() + "}";
+    }
+
+    /**
+     * Awaits {@link #AWAIT_TIMEOUT_MINUTES} minutes for all previously submitted batch-flush tasks to complete.
+     * Restores the interrupted status and exits normally when interrupted during waiting.
+     *
+     * @throws IllegalStateException if tasks did not complete in {@link #AWAIT_TIMEOUT_MINUTES} minutes.
+     */
+    private void awaitCompletion()
+    {
+        try
+        {
+            log.debug( "Waiting " + AWAIT_TIMEOUT_MINUTES + " minutes for all submitted and active " +
+                       "flush tasks to complete." + EOL + this );
+
+            Supplier<Boolean> allSubmittedTasksCompleted = () -> activeTasks.get() == 0;
+            long awaitTimeoutMillis = TimeUnit.MINUTES.toMillis( AWAIT_TIMEOUT_MINUTES );
```

Comment:
too much conversions, you can pass AWAIT_TIMEOUT_MINUTES to await method and timeUnit minutes?